diff --git a/easyshell/src/com/tetrade/eclipse/plugins/easyshell/actions/EasyShellAction.java b/easyshell/src/com/tetrade/eclipse/plugins/easyshell/actions/EasyShellAction.java
index dd02b54..0112d34 100644
--- a/easyshell/src/com/tetrade/eclipse/plugins/easyshell/actions/EasyShellAction.java
+++ b/easyshell/src/com/tetrade/eclipse/plugins/easyshell/actions/EasyShellAction.java
@@ -19,6 +19,7 @@
 package com.tetrade.eclipse.plugins.easyshell.actions;
 
 import java.text.MessageFormat;
+import java.util.StringTokenizer;
 
 import org.eclipse.jface.action.IAction;
 import org.eclipse.jface.dialogs.MessageDialog;
@@ -153,7 +154,14 @@
 					if (ActionIDNum == 3) {
 						cmdAll += cmd;
 					} else {
-						Runtime.getRuntime().exec(cmd);
+						StringTokenizer st = new StringTokenizer(target);
+						String[] cmds = new String[st.countTokens()];
+						int c = 0;
+						while (st.hasMoreElements()) {
+							cmds[c] = MessageFormat.format(st.nextToken(), (Object[])args);
+							c++;
+						}
+						Runtime.getRuntime().exec(cmds);
 					}
 
 				} catch (Exception e) {
diff --git a/easyshell/src/com/tetrade/eclipse/plugins/easyshell/preferences/EasyShellPreferencePage.java b/easyshell/src/com/tetrade/eclipse/plugins/easyshell/preferences/EasyShellPreferencePage.java
index d7e2f1c..8045334 100644
--- a/easyshell/src/com/tetrade/eclipse/plugins/easyshell/preferences/EasyShellPreferencePage.java
+++ b/easyshell/src/com/tetrade/eclipse/plugins/easyshell/preferences/EasyShellPreferencePage.java
@@ -65,7 +65,7 @@
         cmdKonsoleKDEKonqueror(4, "KDE Konsole / Konqueror", 		"konsole --noclose --workdir {1}", 						"konsole --noclose --workdir {1} -e ./{3}", 							"konqueror file:{2}", 			"{2}{5}"),
         cmdKonsoleGnome(5, "Gnome Terminal / Nautilus", 	"gnome-terminal --working-directory={1}", 				"gnome-terminal --working-directory={1} --command=./{3}", 				"nautilus {2}", 				"{2}{5}"),
         cmdXtermDtfile(6, "CDE Xterm / Dtfile", 			"cd {1} && xterm", 										"cd {1} && xterm -e ./{3}", 											"cd {1} && dtfile", 			"{2}{5}"),
-        cmdTerminalFinder(7, "MAC OS X Terminal / Finder",	"open -a Terminal {1}", 								"cd {1} && open {3}",													"open -a Finder {2}",			"{2}{5}"),
+        cmdTerminalFinder(7, "MAC OS X Terminal / Finder",	"open -a Terminal {1}", 								"open -a Terminal {2}",													"open -R {2}",			"{2}{5}"),
         cmdKonsoleKDEDolphin(8, "KDE Konsole / Dolphin", 		"konsole --workdir {1}", 								"konsole --workdir {1} --noclose -e {2}", 								"dolphin --select {2}", 		"{2}{5}");
         // attributes
         private final int id;